### PR TITLE
Fixed #26184 -- Allowed custom lookups and transforms in ModelAdmin.search_fields.

### PR DESCRIPTION
**Pull request by nazarewk**
Adds `!` operator to `search_fields` allowing for specification of custom lookups.

https://code.djangoproject.com/ticket/26184

### CODE REVIEW

**Comment by schinckel** [comment]
I wonder if just returning `field_name`, if it contains `__` (or something similar) would give satisfactory results (meaning you can use arbitrary lookups/transforms).

**Comment by nazarewk** [comment]
@schinckel i'm not sure about others, but for me it is quite common to have stuff like `event__name` in `search_fields`, also my approach is backwards compatible.

**Comment by timgraham** [comment]
I'm not sure if this completely solves the issue. In particular, there's a usage of the legacy [QUERY_TERMS](https://github.com/django/django/blob/61fd2b494a0c5c0b814e36f7032109239cdebbfc/django/contrib/admin/utils.py#L30) which seems like it may crash on custom lookups. `endswith` is a built-in lookup in `QUERY_TERMS` and thus wouldn't reveal a problem, I think.

File: docs/ref/contrib/admin/index.txt [hunk 1]
```diff
@@ -1230,6 +1230,14 @@ subclass::
         default search method but uses an index. Currently this is only
         available for MySQL.
 
+    ``!``
+        Use the '!' operator to specify your own field lookup. For example
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: Add a `.. versionadded:: 2.0` annotation and a mention in `docs/releases/2.0.txt`.  See our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist).

File: docs/ref/contrib/admin/index.txt [hunk 2]
```diff
@@ -1230,6 +1230,14 @@ subclass::
...
+        if ``search_fields`` is set to ``['!first_name__exact']`` and
+        user searches for ``John Lennon``, Django will do the equivalent
+        of this SQL ``WHERE`` clause::
+
+            WHERE (first_name = 'John Lennon')
+
```
Comment (timgraham) [inline docs/ref/contrib/admin/index.txt]: Perhaps how this allows avoiding the concerns in https://code.djangoproject.com/ticket/26001 could also be documented.

Comment (nazarewk) [inline docs/ref/contrib/admin/index.txt]: I am not 100% sure this resolves the issue, but worth giving a hint.

**Comment by timgraham** [comment]
And will you address the `QUERY_TERMS` comment? We need a test for that by registering a custom lookup in a test. Search `with register_lookup` to see existing tests that do that. Here's a draft patch for the fix:
```diff
diff --git a/django/contrib/admin/utils.py b/django/contrib/admin/utils.py
index 5f3a832..a79b798 100644
--- a/django/contrib/admin/utils.py
+++ b/django/contrib/admin/utils.py
@@ -26,21 +26,26 @@ def lookup_needs_distinct(opts, lookup_path):
     Return True if 'distinct()' should be used to query the given lookup path.
     """
     lookup_fields = lookup_path.split(LOOKUP_SEP)
-    # Remove the last item of the lookup path if it is a query term
-    if lookup_fields[-1] in QUERY_TERMS:
-        lookup_fields = lookup_fields[:-1]
-    # Now go through the fields (following all relations) and look for an m2m
+    # Go through the fields (following all relations) and look for an m2m
+    prev_field = None
     for field_name in lookup_fields:
         if field_name == 'pk':
             field_name = opts.pk.name
-        field = opts.get_field(field_name)
-        if hasattr(field, 'get_path_info'):
-            # This field is a relation, update opts to follow the relation
-            path_info = field.get_path_info()
-            opts = path_info[-1].to_opts
-            if any(path.m2m for path in path_info):
-                # This field is a m2m relation so we know we need to call distinct
-                return True
+        try:
+            field = opts.get_field(field_name)
+        except FieldDoesNotExist:
+            # Ignore valid query lookups.
+            if prev_field and prev_field.get_lookup(field_name):
+                continue
+        else:
+            prev_field = field
+            if hasattr(field, 'get_path_info'):
+                # This field is a relation, update opts to follow the relation
+                path_info = field.get_path_info()
+                opts = path_info[-1].to_opts
+                if any(path.m2m for path in path_info):
+                    # This field is a m2m relation so we know we need to call distinct
+                    return True
     return False
 
```

**Comment by nazarewk** [comment]
Working on it.

**Comment by timgraham** [comment]
The postgres-specifc test isn't appropriate as we're not trying to test code in contrib.postgres -- did you understand my suggestion about using `with register_lookup`?

**Comment by nazarewk** [comment]
Thought the postgres json lookup would be the most extreme case, but now that i think about it, it's not exactly the same.

**Comment by timgraham** [comment]
Also, remove `django.db.models.sql.constants.QUERY_TERMS` as this is the last usage of that.

**Comment by timgraham** [comment]
I think it might be possible to use the `field.get_lookup()` logic from `lookup_needs_distinct()` in `construct_search()` to determine if `field_name` ends with a valid lookup. That could eliminate the need for the ! syntax.

**Comment by nazarewk** [comment]
@timgraham what do you think about my proposal?
Maybe delete default `icontains` alltogether, it will break backwards compatibility though.

PS: I am going on holidays, will be back at 7th August, can write better documentation then.

**Comment by timgraham** [comment]
That last commit seems complicated and there's some unrelated refactoring which makes it difficult to follow. My idea was along the lines of this (I ignored your last commit and wrote this patch on top of the commit before):
```diff
diff --git a/django/contrib/admin/options.py b/django/contrib/admin/options.py
index cfd9ee0..a880477 100644
--- a/django/contrib/admin/options.py
+++ b/django/contrib/admin/options.py
@@ -891,9 +891,33 @@ class ModelAdmin(BaseModelAdmin):
         """
         # Apply keyword searches.
         def construct_search(field_name):
-            if field_name.startswith('!'):
-                return field_name[1:]
-            elif field_name.startswith('^'):
+            opts = queryset.model._meta
+            lookup_fields = field_name.split(LOOKUP_SEP)
+            # Go through the fields (following all relations) and look for an m2m
+            prev_field = None
+            for path_part in lookup_fields:
+                if path_part == 'pk':
+                    path_part = opts.pk.name
+                try:
+                    field = opts.get_field(path_part)
+                except FieldDoesNotExist:
+                    # Use valid query lookups.
+                    if prev_field and prev_field.get_lookup(path_part):
+                        return field_name
+                else:
+                    prev_field = field
+                    # The rest of the logic from lookup_needs_distinct() --
+                    # need a test (that spans more relations?) to show if it's
+                    # needed.
+                    # if hasattr(field, 'get_path_info'):
+                        # This field is a relation, update opts to follow the relation
+                        # path_info = field.get_path_info()
+                        # opts = path_info[-1].to_opts
+                        # if any(path.m2m for path in path_info):
+                        #    # This field is a m2m relation so we know we need to call distinct
+                        #    return True
+
+            if field_name.startswith('^'):
                 return "%s__istartswith" % field_name[1:]
             elif field_name.startswith('='):
                 return "%s__iexact" % field_name[1:]
diff --git a/tests/admin_changelist/tests.py b/tests/admin_changelist/tests.py
index 6a9d8aa..1dd529e 100644
--- a/tests/admin_changelist/tests.py
+++ b/tests/admin_changelist/tests.py
@@ -436,7 +436,7 @@ class ChangeListTests(TestCase):
         Concert.objects.create(name='Woodstock', group=band)
 
         m = ConcertAdmin(Concert, custom_site)
-        m.search_fields = ['!group__name__cc']
+        m.search_fields = ['group__name__cc']
 
         try:
             Field.register_lookup(Contains, 'cc')
```
This borrows logic from `lookup_needs_distinct()`. Maybe there's a chance to refactor it into something shared, but I think it would be easier to follow the patch without that refactoring in this PR.

File: tests/admin_changelist/tests.py [hunk 3]
```diff
@@ -429,6 +431,61 @@ def test_pk_in_search_fields(self):
...
+        request = self.factory.get('/concert/', data={SEARCH_VAR: 'woodstock'})
+        cl = ChangeList(request, Concert, *get_changelist_args(m))
+        self.assertEqual(cl.queryset.count(), 0)
+
+    def test_custom_lookup_transform_in_search_fields(self):
+        Band.objects.create(name='The Hype', nr_of_members=10)
+
+        m = BandAdmin(Band, custom_site)
+        m.search_fields = [
+            ('nr_of_members__cgt', int)
```
Comment (timgraham) [inline tests/admin_changelist/tests.py]: I'm not sure if there are lookups where it's not the case, but comparisons such as `Choice.objects.filter(votes__gte='2')` seem to work fine with the value as a string so the "transform" stuff seems unnecessary, at this for the first version of this.

**Comment by timgraham** [comment]
I pushed my proposal in https://github.com/django/django/pull/9357. There might be some refactoring to do as you've proposed but it would be more clear to do that in a separate PR, I think. Thanks for the initial patch and please review my PR if you'd like.

**Pull request closed without merging**